File: server/webapp/WEB-INF/rails.new/app/presenters/api_v3/config/environment_variable_representer.rb
##########################################################################
# Copyright 2016 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV3
  module Config
    class EnvironmentVariableRepresenter < ApiV3::BaseRepresenter
      alias_method :environment_variable, :represented

      error_representer({'encryptedValue' => 'encrypted_value'})

      property :isSecure, as: :secure, writable: false
      property :name, writable: false
      property :value, skip_nil: true, writable: false, exec_context: :decorator
      property :encrypted_value, skip_nil: true, writable: false, exec_context: :decorator
      property :errors, exec_context: :decorator, decorator: ApiV3::Config::ErrorRepresenter, skip_parse: true, skip_render: lambda { |object, options| object.empty? }

      def value
        environment_variable.getValueForDisplay() if environment_variable.isPlain
      end

      def encrypted_value
        environment_variable.getValueForDisplay() if environment_variable.isSecure
      end

      def from_hash(data, options={})
        data = data.with_indifferent_access
        environment_variable.deserialize(data[:name], data[:value], data[:secure].to_bool, data[:encrypted_value])
        environment_variable
      end

    end
  end
end

File: api/api-shared-v6/src/main/java/com/thoughtworks/go/apiv6/shared/representers/EnvironmentVariableRepresenter.java
/*
 * Copyright 2018 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thoughtworks.go.apiv6.shared.representers;

import com.thoughtworks.go.api.base.OutputListWriter;
import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.ErrorGetter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.config.EnvironmentVariableConfig;
import org.bouncycastle.crypto.InvalidCipherTextException;

import java.util.HashMap;
import java.util.List;

public class EnvironmentVariableRepresenter {
    public static void toJSON(OutputListWriter jsonWriter, List<EnvironmentVariableConfig> environmentVariableConfigs) {
        environmentVariableConfigs.forEach(environmentVariableConfig -> {
            jsonWriter.addChild(envVarWriter -> toJSON(envVarWriter, environmentVariableConfig));
        });

    }

    public static void toJSON(OutputWriter jsonWriter, EnvironmentVariableConfig environmentVariableConfig) {
        if (!environmentVariableConfig.errors().isEmpty()) {
            jsonWriter.addChild("errors", errorWriter -> {
                HashMap<String, String> mapping = new HashMap<>();
                mapping.put("encryptedValue", "encrypted_value");
                new ErrorGetter(mapping).toJSON(jsonWriter, environmentVariableConfig);

            });
        }

        jsonWriter.add("secure", environmentVariableConfig.isSecure());
        jsonWriter.add("name", environmentVariableConfig.getName());
        if (environmentVariableConfig.isPlain()) {
            jsonWriter.addIfNotNull("value", environmentVariableConfig.getValueForDisplay());
        }
        if (environmentVariableConfig.isSecure()) {
            jsonWriter.addIfNotNull("encrypted_value", environmentVariableConfig.getValueForDisplay());
        }
    }

    public static EnvironmentVariableConfig fromJSON(JsonReader jsonReader) throws InvalidCipherTextException {
        EnvironmentVariableConfig environmentVariableConfig = new EnvironmentVariableConfig();
        String name = jsonReader.getString("name");
        String value = null, encryptedValue = null;
        Boolean secure = false;
        if (jsonReader.hasJsonObject("value")) {
            value = jsonReader.getString("value");
        }
        if (jsonReader.hasJsonObject("encrypted_value")) {
            encryptedValue = jsonReader.getString("encrypted_value");
        }
        if (jsonReader.hasJsonObject("secure")) {
            secure = jsonReader.optBoolean("secure").get();
        }
        environmentVariableConfig.deserialize(name, value, secure, encryptedValue);
        return environmentVariableConfig;
    }
}
